chemclipse/plugins/org.eclipse.chemclipse.msd.model/src/org/eclipse/chemclipse/msd/model/matrix/ExtractedMatrix.java

```diff
@@ -0,0 +1,158 @@
+ * terms of the Eclipse Public License v1.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v10.html
+ * 
+ * Contributors:
+ * Lorenz Gerber - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.chemclipse.msd.model.matrix;
+
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.eclipse.chemclipse.msd.model.core.AbstractIon;
+import org.eclipse.chemclipse.msd.model.core.IIon;
+import org.eclipse.chemclipse.msd.model.core.IIonBounds;
+import org.eclipse.chemclipse.msd.model.core.IScanMSD;
+import org.eclipse.chemclipse.msd.model.core.selection.IChromatogramSelectionMSD;
+import org.eclipse.chemclipse.msd.model.implementation.Ion;
+
+public class ExtractedMatrix {
+	
+	private IChromatogramSelectionMSD selection;
+	private List<IScanMSD> scans;
+	private int numberOfScans;
+	private int numberOfIons;
+	private int startIon;
+	private int stopIon;
+	private float[][] signal;
+	
+	public ExtractedMatrix(IChromatogramSelectionMSD chromatogramSelection) {
+		this.selection = chromatogramSelection;
+		this.numberOfScans = selection.getStopScan() - selection.getStartScan() + 1;
+		this.scans = extractScans();
+		if(checkHighRes( 10)) {
+			throw new IllegalArgumentException();
+		} else {
+			this.startIon = getMinMz();
+			this.stopIon = getMaxMz();
+			this.numberOfIons = this.stopIon - this.startIon + 1;
+			signal = new float[numberOfScans][this.numberOfIons];
+			List<IIon> currentIons;
+			int numberIonsCurrentScan;
+			for(int scanIndex = 0; scanIndex < numberOfScans; scanIndex++) {
+				currentIons = scans.get(scanIndex).getIons();
+				numberIonsCurrentScan = currentIons.size(); 
+				for(int j = 0; j < numberIonsCurrentScan; j++) {
+					signal[ scanIndex ] [((int) Math.round(currentIons.get(j).getIon() - this.startIon))] =  currentIons.get(j).getAbundance();
+				}
+			}
+		}
+	}
+	
+	private Boolean checkHighRes( int limit ) {
+		IIonBounds bounds;
+		double rangeAbs;
+		
+		for(IScanMSD scan: scans) {
+			bounds = scan.getIonBounds();
+			rangeAbs = bounds.getHighestIon().getIon() - bounds.getLowestIon().getIon();
+			if(rangeAbs + limit < scan.getIons().size()) {
+				return (true);
+			}
+		}
+		
+		return false;
+	}
+	
+	private List<IScanMSD> extractScans() {
+
+		List<IScanMSD> scans;
+		int startRT;
+		int stopRT;
+		startRT = selection.getStartRetentionTime();
+		stopRT = selection.getStopRetentionTime();
+		scans = selection.getChromatogram() //
+				.getScans() //
+				.stream() //
+				.filter(s -> s instanceof IScanMSD) //
+				.map(IScanMSD.class::cast) //
+				.filter(s -> s.getRetentionTime() >= startRT) //
+				.filter(s -> s.getRetentionTime() <= stopRT) //
+				.collect(Collectors.toList()); //
+		return (scans);
+	}
+	
+	private int getMinMz() {
```

Comment:
You could improve performance here by calculating min+max in one call see https://stackoverflow.com/questions/55213611/collect-both-min-and-max-in-one-stream for example.

Maybe you can even check if min/max could be calculated as part of "extractScans", because thats another point where all scans are iterated, so atm you interate the scanns three times.